src/Compilers/Core/Portable/InternalUtilities/OneOrMany.cs

```diff
@@ -212,6 +215,25 @@ public bool SequenceEqual(IEnumerable<T> other, IEqualityComparer<T>? comparer =
             return true;
         }
 
+        /// <summary>
+        /// Similarly to <see cref="ImmutableArray{T}"/>, true means that this instance is equal to <paramref name="other"/>,
+        /// not that that all items in the collection are equal.
+        /// </summary>
+        public bool Equals(OneOrMany<T> other)
+            => HasOneItem ? other.HasOneItem && EqualityComparer<T>.Default.Equals(_one, other._one) : _many.Equals(other._many);
```

Comment:
>HasOneItem ? other.HasOneItem && EqualityComparer<T>.Default.Equals(_one, other._one) : _many.Equals(other._many); [](http://example.com/codeflow?start=15&length=114)

The equality implementation doesn't look consistent. For one item, it is a value equality of the item. For many, it is a reference equality of the underlying array.

Reply:
Rather than implementing equality for the type this way, consider implementing and using custom equality comparer where an equality like that is good enough.

Reply:
The semantics of the equality is the same as semantics of `ImmutableArray<T>.Equals`. Both check that all the fields of the structure are equal.

This is used from generated `Update` methods of bound nodes.

Reply:
> The semantics of the equality is the same as semantics of ```ImmutableArray<T>.Equals```.

It doesn't look this way to me. ```ImmutableArray<T>.Equals``` checks reference equality of the underlying array, this one uses value equality sometimes. I would be more comfortable with using reference equality for all flavors

Reply:
I also would like to suggest to not rely on ```ImmutableArray<T>.Equals``` at all but rather check reference equality of items. Or, if we prefer using value equality, then use value equality of underlying items consistently across one and many scenarios.